### PR TITLE
bugfix - no-use-before-declare - exception for destructuring w/ rename

### PR DESCRIPTION
**Pull request by aervin**
#### PR checklist

- [X] Addresses an existing issue: #3761 
- [X] New feature, bugfix, or enhancement
  - [X] Includes tests
- [x] Documentation update

#### Overview of change:
The rule checks that identifiers have a declaration with a `pos` <= to the reference `pos`, which isn't the case when destructuring w/ a rename. 

```ts
var { x: y } = { x: 43 };
```

#### Is there anything you'd like reviewers to focus on?
Any edge cases with the destructuring syntax that would require additional tests?

#### CHANGELOG.md entry:
[bugfix] [`no-use-before-declare`](https://palantir.github.io/tslint/rules/no-use-before-declare/) Fixes false positives when using the destructuring syntax (#3761)

### CODE REVIEW

File: test/rules/no-use-before-declare/test.ts.lint [hunk 1]
```diff
@@ -68,7 +74,8 @@ export {
 };
 
 var undeclaredA = 42;
-let { undeclaredB } = { undeclaredB: 43 };
+var { undeclaredB } = { undeclaredB: 43 };
+var { undeclaredB: undeclaredA } = { undeclaredB: 43 };
 const [ undeclaredC, [undeclaredD] ] = [ 1, [2] ];
```
Comment (suchanlee) [inline test/rules/no-use-before-declare/test.ts.lint]: Can we add test cases for multiple reassignments and reassignment and non-reassignment in the same destructuring?

**Comment by JoshuaKGoldberg** [comment]
Looks like this was fixed by #3849. 🎉
